src/d3a/models/strategy/home_meter.py

```diff
@@ -0,0 +1,577 @@
+                self.area.next_market, 0) - self.owner.get_path_to_root_fees(), validate=False)
+
+    def _validate_consumption_rates(
+            self, initial_rate, final_rate, energy_rate_change_per_update, fit_to_limit):
+        for time_slot in initial_rate.keys():
+            rate_change = None if fit_to_limit else find_object_of_same_weekday_and_time(
+                energy_rate_change_per_update, time_slot)
+
+            self.validator.validate_rate(
+                initial_buying_rate=initial_rate[time_slot],
+                energy_rate_increase_per_update=rate_change,
+                final_buying_rate=find_object_of_same_weekday_and_time(final_rate, time_slot),
+                fit_to_limit=fit_to_limit)
+
+    def _validate_production_rates(
+            self, initial_rate, final_rate, energy_rate_change_per_update, fit_to_limit):
+        for time_slot in initial_rate.keys():
+            rate_change = None if fit_to_limit else find_object_of_same_weekday_and_time(
+                energy_rate_change_per_update, time_slot)
+
+            self.validator.validate_rate(
+                initial_selling_rate=initial_rate[time_slot],
+                final_selling_rate=find_object_of_same_weekday_and_time(final_rate, time_slot),
+                energy_rate_decrease_per_update=rate_change,
+                fit_to_limit=fit_to_limit)
+
+    def _offer_rate_can_be_accepted(self, offer: Offer, market_slot: Market):
+        """Check if the offer rate is less than what the device wants to pay."""
+        max_affordable_offer_rate = self.bid_update.get_updated_rate(market_slot.time_slot)
+        return (
+            round(offer.energy_rate, DEFAULT_PRECISION)
+            <= max_affordable_offer_rate + FLOATING_POINT_TOLERANCE)
+
+    def _event_tick_consumption(self):
+        for market in self.area.all_markets:
+            # One-sided market (only offers are posted)
+            if ConstSettings.IAASettings.MARKET_TYPE == 1:
+                self._one_sided_market_event_tick(market)
+            # Two-sided markets (both offers and bids are posted)
+            elif ConstSettings.IAASettings.MARKET_TYPE in [2, 3]:
+                # Update the price of existing bids to reflect the new rates
+                self.bid_update.update(market, self)
+
+        # Bid prices have been updated, so we increase the counter of the bid updates
+        self.bid_update.increment_update_counter_all_markets(self)
+
+    def _event_tick_production(self):
+        self.offer_update.update(self)
+        self.offer_update.increment_update_counter_all_markets(self)
+
+    def _one_sided_market_event_tick(self, market, offer=None):
+        """
+        Define the behavior of the device on TICK events in one-sided markets (react to offers).
+        """
+        if not self.state.can_buy_more_energy(market.time_slot):
+            return
+
+        try:
+            if offer is None:
+                if not market.offers:
```

Comment:
I would move this to the start of the method, to exit immediately. Same goes for the `if offer.id not in market.offers` below. This would allow the if / else block to be replaced by this `acceptable_offer = self._find_acceptable_offer(market) if not offer else offer`